src/Castle.Core/DynamicProxy/Generators/MetaMethod.cs

```diff
@@ -101,7 +102,30 @@ internal override void SwitchToExplicitImplementation()
 				Attributes |= MethodAttributes.SpecialName;
 			}
 
-			name = string.Format("{0}.{1}", Method.DeclaringType.Name, Method.Name);
+			var nameBuilder = new StringBuilder();
+			AppendTypeNameTo(nameBuilder, Method.DeclaringType);
+			nameBuilder.Append('.');
+			nameBuilder.Append(Method.Name);
+			name = nameBuilder.ToString();
+		}
+
+		private static void AppendTypeNameTo(StringBuilder stringBuilder, Type type)
+		{
+			stringBuilder.Append(type.Name);
+			if (type.GetTypeInfo().IsGenericType)
+			{
+				stringBuilder.Append('[');
+				var genericTypeArguments = type.GetGenericArguments();
+				for (int i = 0, n = genericTypeArguments.Length; i < n; ++i)
```

Comment:
That's correct. The recursion depth will be proportional to the nesting depth of the (generic) type for which it generates a name. Non-generic types, however, will not trigger any recursion.

(Replacing recursion with an imperative `Stack`-and-loop approach would probably be even worse, since it would likely entail an additional heap allocation that can be avoided with recursion.)